src/components/Table/ContextMenu/index.tsx

```diff
@@ -1,56 +1,28 @@
-import React from "react";
 import _find from "lodash/find";
-import { PopoverProps } from "@mui/material";
-
 import { getFieldProp } from "@src/components/fields";
 import { useProjectContext } from "@src/contexts/ProjectContext";
-
+import useContextMenuAtom from "@src/hooks/useContextMenuAtom";
 import { MenuContents } from "./MenuContent";
 
-export type SelectedCell = {
-  rowIndex: number;
-  colIndex: number;
-};
-
-export type ContextMenuRef = {
-  selectedCell: SelectedCell;
-  setSelectedCell: React.Dispatch<React.SetStateAction<SelectedCell | null>>;
-  anchorEl: HTMLElement | null;
-  setAnchorEl: React.Dispatch<
-    React.SetStateAction<PopoverProps["anchorEl"] | null>
-  >;
-};
-
 export default function ContextMenu() {
-  const { contextMenuRef, tableState }: any = useProjectContext();
-  const [anchorEl, setAnchorEl] = React.useState<any | null>(null);
-  const [selectedCell, setSelectedCell] = React.useState<any | null>();
-  const open = Boolean(anchorEl);
-  const handleClose = () => setAnchorEl(null);
-
-  if (contextMenuRef)
-    contextMenuRef.current = {
-      anchorEl,
-      setAnchorEl,
-      selectedCell,
-      setSelectedCell,
-    } as {};
-
+  const { tableState }: any = useProjectContext();
+  const { contextMenu, resetContextMenu } = useContextMenuAtom();
+  const { anchorEl, selectedCell } = contextMenu;
+  const columns = tableState?.columns;
   const selectedColIndex = selectedCell?.colIndex;
-  const selectedCol = _find(tableState?.columns, { index: selectedColIndex });
-  const getActions =
+  const selectedCol = _find(columns, { index: selectedColIndex });
+  const configActions =
     getFieldProp("contextMenuActions", selectedCol?.type) ||
     function empty() {};
-  const actions = getActions() || [];
+  const actions = configActions(selectedCell, resetContextMenu) || [];
   const hasNoActions = Boolean(actions.length === 0);
 
-  if (!contextMenuRef.current || hasNoActions) return null;
-
+  if (!contextMenu || hasNoActions) return <></>;
   return (
     <MenuContents
-      anchorEl={anchorEl}
-      open={open}
-      handleClose={handleClose}
+      anchorEl={anchorEl as HTMLElement}
```

Comment:
Side note: I think if you add `!anchorEl` as a condition on line 20, TypeScript won’t make you explicitly state `as HTMLElement` here.

Logically we know that if `contextMenu` is falsy, then `contextMenu.anchorEl` would not exist, but since you use `anchorEl` directly, TypeScript isn’t quite smart enough to know that they are the same.